cocotb/share/lib/gpi_log/gpi_logging.c

```diff
@@ -149,50 +159,82 @@ void gpi_log(const char *name, long level, const char *pathname, const char *fun
     // calling args is level, filename, lineno, msg, function
     //
     PyGILState_STATE gstate = PyGILState_Ensure();
+    PyObject *pLogArgs = NULL;
 
-    PyObject *check_args = PyTuple_New(1);
-    PyTuple_SetItem(check_args, 0, PyLong_FromLong(level));
+    PyObject *pCheckArgs = PyTuple_New(1);                  // New reference
+    if (pCheckArgs == NULL)
+        goto error;
 
-    PyObject *filter_ret = PyObject_CallObject(pLogFilter, check_args);
-    Py_DECREF(check_args);
-    if (filter_ret == NULL) {
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
-    }
-    int is_enabled = PyObject_IsTrue(filter_ret);
-    Py_DECREF(filter_ret);
+    PyObject *pLevel = PyLong_FromLong(level);              // New reference
+    if (pLevel == NULL)
+        goto error;
+    if (0 != PyTuple_SetItem(pCheckArgs, 0, pLevel))        // Note: This function steals a reference to pLevel
+        goto error;
+
+    PyObject *pFilterReturn = PyObject_CallObject(pLogFilter, pCheckArgs);
+
+    if (pFilterReturn == NULL)
+        goto error;
+
+    int is_enabled = PyObject_IsTrue(pFilterReturn);
+    Py_DECREF(pFilterReturn);
     if (is_enabled < 0) {
-        /* A python exception occured while converting `filter_ret` to bool */
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
+        /* A python exception occured while converting `pFilterReturn` to bool */
+        goto error;
     }
 
     if (!is_enabled) {
-        PyGILState_Release(gstate);
-        return;
+        goto ok;
     }
 
     va_start(ap, msg);
     n = vsnprintf(log_buff, LOG_SIZE, msg, ap);
     va_end(ap);
 
-    PyObject *call_args = PyTuple_New(5);
-    PyTuple_SetItem(call_args, 0, PyLong_FromLong(level));           // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 1, PyUnicode_FromString(pathname));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 2, PyLong_FromLong(lineno));          // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 3, PyUnicode_FromString(log_buff));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 4, PyUnicode_FromString(funcname));
-
-    PyObject *handler_ret = PyObject_CallObject(pLogHandler, call_args);
-    Py_DECREF(call_args);
```

Comment:
If you're going to change them, you can simplify these lines to
```C
PyObject *handler_ret = PyObject_CallFunction(pLogHandler, "lslss", level, pathname, lineno, log_buff, funcname);
```
which handles all the ref-counting for you.

Reply:
Crap, that doesn't mean the right thing on python 2 :(. I think what you have is the only option

Reply:
Are you sure? `PyObject_CallFunction()` looks like it fits this case in the Python 2.7 docs.

Reply:
It works locally. The 2.7 build on linux is failing because of a const warning elevated to error.
Passing in a string literal to the `PyObject_CallFuction()` format argument causes a const mismatch...
Python 3.4+ changed to a const char * so it builds fine.

Reply:
Added `#ifdef`s for Python 2/3 to get it building without warnings.

Reply:
Now it has `gpi_logging.c:166:65: error: cast discards 'const' qualifier from pointer target type [-Werror=cast-qual]`
I thought this might happen. Is there an easy way to make gcc happy here for 2.7?

Reply:
Aide from compilation, the problem is that "s" will construct a `str` object in Python 2, but the current code constructs `unicode` objects.

Reply:
Using `const_cast<char *>("...")` may solve the warning.

Reply:
Perhaps worth abandoning this rabbit home for now.

Reply:
```python
char fmt[] = "lslss";
PyObject *handler_ret = PyObject_CallFunction(pLogHandler, fmt, level, pathname, lineno, log_buff, funcname);
```
Works.

Reply:
Ok, I split the difference. Still creating `unicode` objects but calling with `PyObject_CallFunctionObjArgs()` so we don't have to build the tuple first.

I went with @ktbarrett's method, but another option would be adding `-Wno-discarded-qualifiers` to the Makefile.

Reply:
I'd thought this was a C++ file. Evidently not!

Reply:
`const_cast` is never a good idea. I have no idea what Python does with the builder string, but if it tries to modify it in a function and you give it a `const_cast` string literal (string literals are stored in read-only section of the object), you'll end up with a crash. Likely, Python isn't modifying it and just forgot the qualifier, hooray.

Reply:
@eric-wieser Is there any reason why all of the sources aren't C++ files?

Reply:
@ktbarrett:

> Likely, Python isn't modifying it and just forgot the qualifier, hooray.

If you can prove this to be the case by reading the source code, then `const_cast` should be perfectly fine. I believe this is exactly what happened - it comes from a time when compilers were less strict about pointer types.